### PR TITLE
feat(sidenav): added onClose callback

### LINKED ISSUE #3179: Sidenav : Is there a way to call a callback on escape or backdrop close ?
Hi,
I have an ui-view in my sidenav and I'd like to reset the router path when the user click outside the sidenav (on the backdrop) or press escape.
Is there a way to do so ?
Thanks for your help

### PR DESCRIPTION
**Pull request by EladBezalel**
Added onClose callback to handle close, backdrop click or escape key pressed

fixes #3179

### CODE REVIEW

File: src/components/sidenav/sidenav.js [hunk 1]
```diff
@@ -55,6 +55,10 @@ angular
  * // Sync check to whether given sidenav is locked open
  * // If this is true, the sidenav will be open regardless of close()
  * $mdSidenav(componentId).isLockedOpen();
+ * // On close callback to handle close, backdrop click or escape key pressed
```
Comment (ThomasBurleson) [inline src/components/sidenav/sidenav.js]: Please note that the callback happens BEFORE the close action occurs.
Should we also support/allow a callback return value to "prevent" the close intention?

Comment (EladBezalel) [inline src/components/sidenav/sidenav.js]: We should, it's called mdIsLockedOpen

**Comment by ThomasBurleson** [comment]
@EladBezalel - needs rebased with latest changes to `$mdSidenav` service.

File: src/components/sidenav/sidenav.js [hunk 2]
```diff
@@ -79,6 +84,10 @@ function SidenavService($mdComponentRegistry, $q) {
       isLockedOpen: function() {
         return instance && instance.isLockedOpen();
       },
+      onClose: function (callback) {
+        if (instance) instance.onClose = callback;
```
Comment (ThomasBurleson) [inline src/components/sidenav/sidenav.js]:
```js
if (instance) instance.onClose = callback || angular.noop;
```

**Comment by gkohen** [comment]
Team, this is not a P4 or a since to have in my opinion. There is no good way to have a logic run when the sidenav closes as a result of clicking on the backdrop.

**Comment by topherfangio** [comment]
@EladBezalel Can you rebase this when you have a minute?

@gkohen Are you saying that it should be a higher priority?

**Comment by gkohen** [comment]
@topherfangio, I'd say the answer is: 'yes' since it's an important enough functionality of which only workaround is polling for isClose.

**Comment by EladBezalel** [comment]
@topherfangio updated :)

**Comment by topherfangio** [comment]
@ThomasBurleson LGTM 👍

**Comment by ptitdje45** [comment]
```
 $mdSidenav('right').onClose(function() {
    //blabla
      });
    });
```

A little problem about it : "SideNav 'right' is not available! Did you use md-component-id='right'?"

**Pull request closed without merging**